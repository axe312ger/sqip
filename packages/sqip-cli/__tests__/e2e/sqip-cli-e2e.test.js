import { resolve } from 'path'
import { tmpdir } from 'os'
import { stat, remove, readFile } from 'fs-extra'

import cheerio from 'cheerio'
import execa from 'execa'

const inputFile = resolve(
  __dirname,
  '..',
  '..',
  '..',
  '..',
  '__tests__',
  'fixtures',
  'beach.jpg'
)
const cliPath = resolve(__dirname, '..', '..', 'dist', 'wrapper.js')
const cliCmd = `node`

jest.setTimeout(20000)

function isValidStdout(stdout) {
  expect(stdout).toMatch(/Processing: \/([A-z0-9-_+]+\/)*[A-z0-9-_]+\.jpg/)
  expect(stdout).toMatch(/Stored at: \/([A-z0-9-_+]+\/)*[A-z0-9-_]+\.svg/)
  expect(stdout).toMatch(/originalWidth.+originalHeight.+width.+height.+type/)
  expect(stdout).toMatch(
    /Vibrant.+DarkVibrant.+LightVibrant.+Muted.+DarkMuted.+LightMuted/
  )
}

describe('cli api', () => {
  test('no config exists programm and shows help', async () => {
    try {
      await execa(cliCmd, [cliPath], {
        stripFinalNewline: true
      })
      throw new Error('cli should exit with help message')
    } catch (err) {
      expect(err.stdout).toMatchSnapshot()
      expect(err.stderr).toMatchSnapshot()
      expect(err.code).toBe(1)
    }
  })
  test('--help shows help screen to user', async () => {
    const { stdout } = await execa(cliCmd, [cliPath, '--help'], {
      stripFinalNewline: true
    })
    expect(stdout).toMatchSnapshot()
  })
  test('no output will not show stored at hint', async () => {
    const { stdout } = await execa(cliCmd, [cliPath, '--input', inputFile, '-n', 3], {
      stripFinalNewline: true
    })

    expect(stdout).not.toMatch(/Stored at:/)
  })
  test('--silent disables logging to stdout', async () => {
    const { stdout } = await execa(
      cliCmd,
      [cliPath, '--input', inputFile, '--silent', '-n', 3],
      {
        stripFinalNewline: true
      }
    )

    expect(stdout).toMatch("")
  })
  test('-o save result to file and basic svg structure is applied', async () => {
    const outputFile = resolve(
      tmpdir(),
      `sqip-e2e-test-${new Date().getTime()}.svg`
    )

    const { stdout } = await execa(
      cliCmd,
      [cliPath, '-i', inputFile, '-o', outputFile, '-n', 3],
      {
        stripFinalNewline: true
      }
    )
    isValidStdout(stdout)

    // Does the new file exist
    expect(await stat(outputFile)).toBeTruthy()

    const content = await readFile(outputFile)
    const $ = cheerio.load(content, { xml: true })

    // File content is actually a parseable svg
    expect($('svg')).toHaveLength(1)

    // Check default blur value
    const $filter = $('svg > filter > feGaussianBlur')
    expect($filter).toHaveLength(1)
    expect($filter.attr('stdDeviation')).toBe('12')

    await remove(outputFile)
  })
})
