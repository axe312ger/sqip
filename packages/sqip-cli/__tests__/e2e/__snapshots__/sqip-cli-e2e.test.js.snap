// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cli api --help shows help screen to user 1`] = `
"
Usage: sqip --input [path]

Options:
  --version                 Show version number                        [boolean]
  --input, -i                                                [string] [required]
  --output, -o              Save the resulting SVG to a file. The svg result
                            will be returned by default.                [string]
  --numberOfPrimitives, -n  The number of primitive shapes to use to build the
                            SQIP SVG                       [number] [default: 8]
  --mode, -m                The style of primitives to use:
                            0=combo, 1=triangle, 2=rect, 3=ellipse, 4=circle,
                            5=rotatedrect, 6=beziers, 7=rotatedellipse,
                            8=polygon                      [number] [default: 0]
  --blur, -b                Set the GaussianBlur SVG filter value. Disable it
                            via 0.                        [number] [default: 12]

Examples:
  sqip --input /path/to/input.jpg           Output input.jpg image as SQIP
  sqip -i input.jpg -n 25 -b 0 -o           Save input.jpg as result.svg with 25
  result.svg                                shapes and no blur

\\"SQIP\\" (pronounced \\\\skwɪb\\\\ like the non-magical folk of magical descent) is a
SVG-based LQIP technique - https://github.com/technopagan/sqip
"
`;

exports[`cli api no config exists programm and shows help 1`] = `
"
Usage: sqip --input [path]

Options:
  --help                    Show help                                  [boolean]
  --version                 Show version number                        [boolean]
  --input, -i                                                [string] [required]
  --output, -o              Save the resulting SVG to a file. The svg result
                            will be returned by default.                [string]
  --numberOfPrimitives, -n  The number of primitive shapes to use to build the
                            SQIP SVG                       [number] [default: 8]
  --mode, -m                The style of primitives to use:
                            0=combo, 1=triangle, 2=rect, 3=ellipse, 4=circle,
                            5=rotatedrect, 6=beziers, 7=rotatedellipse,
                            8=polygon                      [number] [default: 0]
  --blur, -b                Set the GaussianBlur SVG filter value. Disable it
                            via 0.                        [number] [default: 12]

Examples:
  sqip --input /path/to/input.jpg           Output input.jpg image as SQIP
  sqip -i input.jpg -n 25 -b 0 -o           Save input.jpg as result.svg with 25
  result.svg                                shapes and no blur

\\"SQIP\\" (pronounced \\\\skwɪb\\\\ like the non-magical folk of magical descent) is a
SVG-based LQIP technique - https://github.com/technopagan/sqip

Missing required argument: input"
`;
